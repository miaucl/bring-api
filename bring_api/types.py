"""Bring API types."""
from enum import Enum
from typing import List, TypedDict


class BringList(TypedDict):
    """A list class. Represents a single list."""

    listUuid: str
    name: str
    theme: str


class BringPurchase(TypedDict):
    """A purchase class. Represents a single item."""

    uuid: str
    itemId: str
    specification: str


class BringListItemDetails(TypedDict):
    """An item details class.

    Includes several details of an item in the context of a list.
    Caution: This does not have to be an item that is currently marked as 'to buy'.
    """

    uuid: str
    itemId: str
    listUuid: str
    userIconItemId: str
    userSectionId: str
    assignedTo: str
    imageUrl: str


class BringAuthResponse(TypedDict):
    """An auth response class."""

    uuid: str
    publicUuid: str
    email: str
    name: str
    photoPath: str
    bringListUUID: str
    access_token: str
    refresh_token: str
    token_type: str
    expires_in: int


class BringListResponse(TypedDict):
    """A list response class."""

    lists: List[BringList]


class BringItemsResponse(TypedDict):
    """An items response class."""

    uuid: str
    status: str
    purchase: List[BringPurchase]
    recently: List[BringPurchase]


class BringListItemsDetailsResponse(List[BringListItemDetails]):
    """A response class of a list of item details."""


class BringNotificationType(Enum):
    """Notification type.

    GOING_SHOPPING: "I'm going shopping! - Last chance for adjustments"
    CHANGED_LIST: "List changed - Check it out"
    SHOPPING_DONE: "Shopping done - you can relax"
    URGENT_MESSAGE: "Breaking news - Please get {itemName}!
    """

    GOING_SHOPPING = "GOING_SHOPPING"
    CHANGED_LIST = "CHANGED_LIST"
    SHOPPING_DONE = "SHOPPING_DONE"
    URGENT_MESSAGE = "URGENT_MESSAGE"


class BringNotificationsConfigType(TypedDict):
    """A notification config."""

    arguments: List[str]
    listNotificationType: str
    senderPublicUserUuid: str


class BringUserSettingsEntry(TypedDict):
    """A user settings class. Represents a single user setting."""

    key: str
    value: str


class BringUserListSettingEntry(TypedDict):
    """A user list settings class. Represents a single list setting."""

    listUuid: str
    usersettings: List[BringUserSettingsEntry]


class BringUserSettingsResponse(TypedDict):
    """A user settings response class."""

    usersettings: List[BringUserSettingsEntry]
    userlistsettings: List[BringUserListSettingEntry]


class BringSyncCurrentUserResponse(TypedDict):
    """A sync current user response class."""

    email: str
    name: str
    publicUuid: str
    userLocale: dict[str, str]
    userUuid: str
