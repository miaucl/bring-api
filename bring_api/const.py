"""Constants for bring-api."""
from typing import Final

API_BASE_URL: Final = "https://api.getbring.com/rest/"
DEFAULT_HEADERS: Final = {
    "Authorization": "Bearer",
    "X-BRING-API-KEY": "cof4Nc6D8saplXjE3h3HXqHH8m7VU2i1Gs0g85Sp",
    "X-BRING-CLIENT": "android",
    "X-BRING-APPLICATION": "bring",
    "X-BRING-COUNTRY": "DE",
    "X-BRING-USER-UUID": "",
}

LOCALES_BASE_URL: Final = "https://web.getbring.com/locale/"
BRING_SUPPORTED_LOCALES: Final = [
    "en-AU",
    "de-DE",
    "fr-FR",
    "it-IT",
    "en-CA",
    "nl-NL",
    "nb-NO",
    "pl-PL",
    "pt-BR",
    "ru-RU",
    "sv-SE",
    "de-CH",
    "fr-CH",
    "it-CH",
    "es-ES",
    "tr-TR",
    "en-GB",
    "en-US",
    "hu-HU",
    "de-AT",
]

BRING_DEFAULT_LOCALE: Final = "de-CH"
