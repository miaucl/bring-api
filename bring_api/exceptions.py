"""Bring API exceptions."""


class BringException(Exception):
    """General exception occurred."""


class BringAuthException(BringException):
    """When an authentication error is encountered."""


class BringRequestException(BringException):
    """When the HTTP request fails."""


class BringParseException(BringException):
    """When parsing the response of a request fails."""


class BringEMailInvalidException(BringException):
    """When checkemail returns emailValid false ."""


class BringUserUnknownException(BringException):
    """When checkemail returns userExists false."""


class BringTranslationException(BringException):
    """When translating an article fails."""
