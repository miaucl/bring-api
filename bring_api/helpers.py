"""Helpers for the Bring API."""

import json
from typing import cast


def headers_serialize(headers: dict[str, str]) -> str:
    """Serialize all headers from the login session to a string for persistency.

    Parameters
    ----------
    headers
        The headers in the session to persist

    str
        The serialized data to persist

    """
    return json.dumps(headers)


def headers_deserialize(headers: str) -> dict[str, str]:
    """Deserialize all headers from the login session from a persisted string.

    Parameters
    ----------
    headers
        The persisted session headers as string

    Returns
    -------
    dict[str, str]
        The session cookies

    """
    return cast(dict[str, str], json.loads(headers))
