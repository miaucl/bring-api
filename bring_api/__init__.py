"""Bring API package."""

__VERSION__ = "0.8.1a1"

from .bring import Bring
from .exceptions import (
    BringAuthException,
    BringEMailInvalidException,
    BringParseException,
    BringRequestException,
    BringTranslationException,
    BringUserUnknownException,
)
from .types import (
    BringAuthResponse,
    BringAuthTokenRespone,
    BringItem,
    BringItemOperation,
    BringItemsResponse,
    BringListItemDetails,
    BringListItemsDetailsResponse,
    BringListResponse,
    BringNotificationsConfigType,
    BringNotificationType,
    BringSyncCurrentUserResponse,
    BringUserListSettingEntry,
    BringUserSettingsEntry,
    BringUserSettingsResponse,
)

__all__ = [
    "Bring",
    "BringAuthException",
    "BringAuthResponse",
    "BringAuthTokenRespone",
    "BringEMailInvalidException",
    "BringItem",
    "BringItemOperation",
    "BringItemsResponse",
    "BringListItemDetails",
    "BringListItemsDetailsResponse",
    "BringListResponse",
    "BringNotificationsConfigType",
    "BringNotificationType",
    "BringParseException",
    "BringRequestException",
    "BringSyncCurrentUserResponse",
    "BringTranslationException",
    "BringUserListSettingEntry",
    "BringUserSettingsEntry",
    "BringUserSettingsResponse",
    "BringUserUnknownException",
]
