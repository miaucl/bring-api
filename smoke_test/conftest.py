"""Smoke test for bring-api."""

import logging
import os

import aiohttp
from dotenv import load_dotenv
import pytest

from bring_api.bring import Bring
from bring_api.helpers import headers_deserialize, headers_serialize
from bring_api.types import BringList

load_dotenv()

_LOGGER = logging.getLogger(__name__)


def save_headers(headers: dict[str, str]) -> None:
    """Save the headers locally."""
    with open(".headers", "w", encoding="utf-8") as file:
        file.write(headers_serialize(headers))


def load_headers() -> dict[str, str]:
    """Load the headers locally."""
    # Open and read the file
    with open(".headers", encoding="utf-8") as file:
        return headers_deserialize(file.read() or "{}")


@pytest.fixture(name="headers")
async def cookies_str() -> dict[str, str]:
    """Load the headers."""

    return load_headers()


@pytest.fixture(name="session")
async def aiohttp_client_session():
    """Create  a client session."""
    async with aiohttp.ClientSession() as session:
        yield session


@pytest.fixture(name="bring")
async def bring_api_client(
    session: aiohttp.ClientSession, headers: dict[str, str]
) -> Bring:
    """Create Bring instance."""

    bring = Bring(session, os.environ["EMAIL"], os.environ["PASSWORD"])

    # Restore auth data from saved headers
    bring.headers = headers
    bring.uuid = headers["X-BRING-USER-UUID"]
    bring.public_uuid = headers["X-BRING-PUBLIC-USER-UUID"]
    await bring.reload_user_list_settings()
    await bring.reload_article_translations()

    return bring


@pytest.fixture(name="bring_no_auth")
async def bring_no_auth_api_client(session: aiohttp.ClientSession) -> Bring:
    """Create Bring instance."""

    bring = Bring(session, os.environ["EMAIL"], os.environ["PASSWORD"])

    return bring


@pytest.fixture(name="test_list")
async def test_list(bring: Bring) -> BringList:
    """Get the BringList instance to test with."""
    # Get information about all available shopping list and select the one to test with
    lists = (await bring.load_lists()).lists
    lst = next(lst for lst in lists if lst.name == os.environ["LIST"])
    _LOGGER.info("Selected list: %s (%s)", lst.name, lst.listUuid)
    return lst
