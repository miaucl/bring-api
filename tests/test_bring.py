"""Unit tests for bring-api."""

import asyncio
import enum
from unittest.mock import patch

import aiohttp
from dotenv import load_dotenv
import pytest

from bring_api.exceptions import (
    BringAuthException,
    BringEMailInvalidException,
    BringRequestException,
    BringUserUnknownException,
)
from bring_api.types import BringNotificationType

from .conftest import (
    BRING_GET_LISTS_RESPONSE,
    BRING_LOGIN_RESPONSE,
    BRING_USER_ACCOUNT_RESPONSE,
    BRING_USER_SETTINGS_RESPONSE,
    UUID,
)

load_dotenv()


class TestDoesUserExist:
    """Tests for does_user_exist method."""

    async def test_mail_invalid(self, mocked, bring):
        """Test does_user_exist for invalid e-mail."""
        mocked.get("https://api.getbring.com/rest/bringusers?email=EMAIL", status=400)
        with pytest.raises(BringEMailInvalidException):
            await bring.does_user_exist("EMAIL")

    async def test_unknown_user(self, mocked, bring):
        """Test does_user_exist for unknown user."""
        mocked.get("https://api.getbring.com/rest/bringusers?email=EMAIL", status=404)
        with pytest.raises(BringUserUnknownException):
            await bring.does_user_exist("EMAIL")

    async def test_user_exist_with_parameter(self, mocked, bring):
        """Test does_user_exist for known user."""
        mocked.get("https://api.getbring.com/rest/bringusers?email=EMAIL", status=200)
        assert await bring.does_user_exist("EMAIL") is True

    async def test_user_exist_without_parameter(self, mocked, bring):
        """Test does_user_exist for known user."""
        mocked.get(
            "https://api.getbring.com/rest/bringusers?email=EMAIL",
            status=200,
        )
        assert await bring.does_user_exist() is True


class TestLogin:
    """Tests for login method."""

    async def test_mail_invalid(self, mocked, bring):
        """Test login with invalid e-mail."""
        mocked.post(
            "https://api.getbring.com/rest/v2/bringauth",
            status=400,
        )
        with pytest.raises(
            BringAuthException,
            match="Login failed due to bad request, please check your email.",
        ):
            await bring.login()

    async def test_unauthorized(self, mocked, bring):
        """Test login with unauthorized user."""
        mocked.post(
            "https://api.getbring.com/rest/v2/bringauth",
            status=401,
            payload={"message": ""},
        )
        with pytest.raises(
            BringAuthException,
            match="Login failed due to authorization failure, "
            "please check your email and password.",
        ):
            await bring.login()

    @pytest.mark.parametrize(
        "exception",
        [
            asyncio.TimeoutError,
            aiohttp.ClientError,
        ],
    )
    async def test_exceptions(self, mocked, bring, exception):
        """Test exceptions."""
        mocked.post("https://api.getbring.com/rest/v2/bringauth", exception=exception)
        with pytest.raises(BringRequestException):
            await bring.login()

    async def test_successfull_login(self, mocked, bring):
        """Test login with valid user."""

        mocked.post(
            "https://api.getbring.com/rest/v2/bringauth",
            status=200,
            payload=BRING_LOGIN_RESPONSE,
        )
        mocked.get(
            f"https://api.getbring.com/rest/v2/bringusers/{UUID}",
            status=200,
            payload=BRING_USER_ACCOUNT_RESPONSE,
        )
        mocked.get(
            f"https://api.getbring.com/rest/bringusersettings/{UUID}",
            status=200,
            payload=BRING_USER_SETTINGS_RESPONSE,
        )

        await bring.login()

        assert bring.headers["Authorization"] == "Bearer ACCESS_TOKEN"
        assert bring.uuid == UUID
        assert bring.public_uuid == UUID
        assert bring.user_locale == "de-DE"
        assert bring.user_list_settings[UUID]["listArticleLanguage"] == "de-DE"


async def test_load_lists(bring, mocked):
    """Test load_lists."""

    mocked.get(
        f"https://api.getbring.com/rest/bringusers/{UUID}/lists",
        status=200,
        payload=BRING_GET_LISTS_RESPONSE,
    )
    with patch.object(bring, "uuid", UUID):
        lists = await bring.load_lists()
        assert isinstance(lists, dict) is True, "load_lists must return a TypedDict"
        assert len(lists["lists"]) > 0, "lists count is 0"


class TestNotifications:
    """Tests for notification method."""

    @pytest.mark.parametrize(
        ("notification_type", "item_name"),
        [
            (BringNotificationType.GOING_SHOPPING, ""),
            (BringNotificationType.CHANGED_LIST, ""),
            (BringNotificationType.SHOPPING_DONE, ""),
            (BringNotificationType.URGENT_MESSAGE, "WITH_ITEM_NAME"),
        ],
    )
    async def test_notify(
        self,
        bring,
        notification_type: BringNotificationType,
        item_name: str,
        mocked,
    ):
        """Test GOING_SHOPPING notification."""

        mocked.post(
            f"https://api.getbring.com/rest/v2/bringnotifications/lists/{UUID}",
            status=200,
        )
        resp = await bring.notify(UUID, notification_type, item_name)
        assert resp.status == 200

    async def test_notify_urgent_message_item_name_missing(self, bring, mocked):
        """Test URGENT_MESSAGE notification."""
        mocked.post(
            f"https://api.getbring.com/rest/v2/bringnotifications/lists/{UUID}",
            status=200,
        )
        with pytest.raises(
            ValueError,
            match="notificationType is URGENT_MESSAGE but argument itemName missing.",
        ):
            await bring.notify(UUID, BringNotificationType.URGENT_MESSAGE, "")

    async def test_notify_notification_type_raise_attribute_error(self, bring, mocked):
        """Test URGENT_MESSAGE notification."""

        with pytest.raises(
            AttributeError,
        ):
            await bring.notify(UUID, "STRING", "")

    async def test_notify_notification_type_raise_type_error(self, bring, mocked):
        """Test URGENT_MESSAGE notification."""

        class WrongEnum(enum.Enum):
            """Test Enum."""

            UNKNOWN = "UNKNOWN"

        with pytest.raises(
            TypeError,
            match="notificationType WrongEnum.UNKNOWN not supported,"
            "must be of type BringNotificationType.",
        ):
            await bring.notify(UUID, WrongEnum.UNKNOWN, "")
