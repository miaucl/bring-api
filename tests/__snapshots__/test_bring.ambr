# serializer version: 1
# name: TestGetActivity.test_get_activity
  BringActivityResponse(timeline=[Activity(type=<ActivityType.LIST_ITEMS_CHANGED: 'LIST_ITEMS_CHANGED'>, content=ActivityContent(uuid='673594a9-f92d-4cb6-adf1-d2f7a83207a4', sessionDate=datetime.datetime(2025, 1, 1, 3, 9, 33, 36000, tzinfo=datetime.timezone.utc), publicUserUuid='98615d7e-0a7d-4a7e-8f73-a9cbb9f1bc32', items=[], purchase=[BringPurchase(uuid='658a3770-1a03-4ee0-94a6-10362a642377', itemId='Gurke', specification='', attributes=[])], recently=[BringPurchase(uuid='1ed22d3d-f19b-4530-a518-19872da3fd3e', itemId='Milch', specification='', attributes=[])])), Activity(type=<ActivityType.LIST_ITEMS_ADDED: 'LIST_ITEMS_ADDED'>, content=ActivityContent(uuid='9a16635c-dea2-4e00-904a-c5034f9cfecf', sessionDate=datetime.datetime(2025, 1, 1, 2, 54, 57, 656000, tzinfo=datetime.timezone.utc), publicUserUuid='6743a171-247d-46d0-bc06-baf31194f949', items=[BringPurchase(uuid='66a633a2-ae09-47bf-8845-3c0198480544', itemId='Joghurt', specification='', attributes=[])], purchase=[], recently=[])), Activity(type=<ActivityType.LIST_ITEMS_REMOVED: 'LIST_ITEMS_REMOVED'>, content=ActivityContent(uuid='303dedf6-d4b2-4d25-a8cd-1c7967b84fcb', sessionDate=datetime.datetime(2025, 1, 1, 3, 9, 12, 380000, tzinfo=datetime.timezone.utc), publicUserUuid='6d79d10b-70b2-443f-9f7e-0b02e670c402', items=[BringPurchase(uuid='2ba8ddb6-01c6-4b0b-a89d-f3da6b291528', itemId='Tofu', specification='', attributes=[])], purchase=[], recently=[]))], timestamp=datetime.datetime(2025, 1, 1, 3, 9, 33, 36000, tzinfo=datetime.timezone.utc), totalEvents=2)
# ---
# name: TestGetAllItemDetails.test_get_all_item_details
  BringListItemsDetailsResponse(items=[BringListItemDetails(uuid='bfb5634c-d219-4d66-b68e-1388e54f0bb0', itemId='Milchreis', listUuid='00000000-0000-0000-0000-000000000000', userIconItemId='Reis', userSectionId='Getreideprodukte', assignedTo='', imageUrl=''), BringListItemDetails(uuid='0056b23c-7fc3-44da-8c34-426f8b632220', itemId='Zitronensaft', listUuid='00000000-0000-0000-0000-000000000000', userIconItemId='Zitrone', userSectionId='Zutaten & Gewürze', assignedTo='', imageUrl='')])
# ---
# name: TestGetAllUserSettings.test_get_all_user_settings
  BringUserSettingsResponse(usersettings=[BringUserSettingsEntry(key='autoPush', value='ON'), BringUserSettingsEntry(key='purchaseStyle', value='grouped'), BringUserSettingsEntry(key='premiumHideSponsoredCategories', value='OFF'), BringUserSettingsEntry(key='premiumHideInspirationsBadge', value='OFF'), BringUserSettingsEntry(key='premiumHideOffersBadge', value='OFF'), BringUserSettingsEntry(key='premiumHideOffersOnMain', value='OFF'), BringUserSettingsEntry(key='defaultListUUID', value='xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxx'), BringUserSettingsEntry(key='discountActivatorOnMainEnabled', value='OFF'), BringUserSettingsEntry(key='onboardClient', value='android')], userlistsettings=[BringUserListSettingEntry(listUuid='00000000-0000-0000-0000-000000000000', usersettings=[BringUserSettingsEntry(key='listSectionOrder', value='["Früchte & Gemüse","Brot & Gebäck","Milch & Käse","Fleisch & Fisch","Zutaten & Gewürze","Fertig- & Tiefkühlprodukte","Getreideprodukte","Snacks & Süsswaren","Getränke & Tabak","Haushalt & Gesundheit","Pflege & Gesundheit","Tierbedarf","Baumarkt & Garten","Eigene Artikel"]'), BringUserSettingsEntry(key='listArticleLanguage', value='de-DE')])])
# ---
# name: TestGetList.test_get_list
  BringItemsResponse(uuid='00000000-0000-0000-0000-000000000000', status=<Status.SHARED: 'SHARED'>, items=Items(purchase=[BringPurchase(uuid='43bdd5a2-740a-4230-8b27-d0bbde886da7', itemId='Paprika', specification='grün', attributes=[]), BringPurchase(uuid='2de9d1c0-c211-4129-b6c5-c1260c3fc735', itemId='Zucchetti', specification='gelb', attributes=[])], recently=[BringPurchase(uuid='5681ed79-c8e4-4c8b-95ec-112999d016c0', itemId='Paprika', specification='rot', attributes=[]), BringPurchase(uuid='01eea2cd-f433-4263-ad08-3d71317c4298', itemId='Pouletbrüstli', specification='', attributes=[])]))
# ---
# name: TestGetListUsers.test_get_lists_users
  BringUsersResponse(users=[BringUser(publicUuid='98615d7e-0a7d-4a7e-8f73-a9cbb9f1bc32', pushEnabled=True, plusTryOut=False, country='DE', language='de', name='NAME', email='EMAIL', photoPath=''), BringUser(publicUuid='73af455f-c158-4004-a5e0-79f4f8a6d4bd', pushEnabled=True, plusTryOut=False, country='US', language='en', name='NAME', email='EMAIL', photoPath=''), BringUser(publicUuid='7d5e9d08-877a-4c36-8740-a9bf74ec690a', pushEnabled=True, plusTryOut=False, country='US', language='en', name=None, email=None, photoPath=None)])
# ---
# name: TestGetUserAccount.test_get_user_account
  BringSyncCurrentUserResponse(email='{email}', emailVerified=True, premiumConfiguration={'hasPremium': False, 'hideSponsoredProducts': False, 'hideSponsoredTemplates': False, 'hideSponsoredPosts': False, 'hideSponsoredCategories': False, 'hideOffersOnMain': False}, publicUserUuid='00000000-0000-0000-0000-000000000000', userLocale=UserLocale(language='de', country='DE'), userUuid='00000000-0000-0000-0000-000000000000', name='{user_name}', photoPath='bring/user/portrait/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxx')
# ---
# name: TestLoadLists.test_load_lists
  BringListResponse(lists=[BringList(listUuid='00000000-0000-0000-0000-000000000000', name='Einkauf', theme='ch.publisheria.bring.theme.home')])
# ---
# name: TestLogin.test_login
  BringAuthResponse(uuid='00000000-0000-0000-0000-000000000000', publicUuid='00000000-0000-0000-0000-000000000000', bringListUUID='00000000-0000-0000-0000-000000000000', access_token='ACCESS_TOKEN', refresh_token='REFRESH_TOKEN', token_type='Bearer', expires_in=604799, photoPath='', email='EMAIL', name='NAME')
# ---
# name: TestRetrieveNewAccessToken.test_retrieve_new_access_token
  BringAuthTokenResponse(access_token='{access_token}', refresh_token='{refresh_token}', token_type='Bearer', expires_in=604799)
# ---
